### DISCUSSION TITLE
Separate units and ingredients, add scaling

### DISCUSSION THREAD

**Discussion by TeddOravec** [mealie-recipes/mealie#694] [Feature Request]
There could be a scale option, default to 1x, that could change how much of each ingredient is needed. For example, instead of 

```
Ingredients:
"1 cup flour",
"1/2 cup butter"
```

It could be
```
Scale: 1.0
Ingredients: [
  {  Name: "butter",
      Amount: 0.5,
      Unit: "cup",
      Plural: "cups",
      Extra: "melted"
  },
Etc
] 
```
Which displays as 

`{{Scale*amount}} {{unit(if >=2, plural}} {{name}} ({{extra}})`
i.e.
`1/2 cup butter (melted)`

**Comment by dpieski** [comment]
Some comments here: 
- Plurality can be more complicated than that. 
If it is greater than 1 it should be plural so `{{unit(if >1, plural}}`
If the unit is in decimals, then if it is less than 1, use plural (e.g., 0.5 apples)
If the unit is in fractions, then if it is less than 1, use singular (e.g., 1/2 apple)
- How do we decide to use factions or decimals? 
- Using something like flour instead of butter, `{{extra}}` could be insufficient. 
For example, `1 cup sifted flour` and `1 cup flour, sifted` actually mean two different things in a recipe. The first means that you sift the flour into a 1 cup measuring cup to use, whereas the second means your take one cup of flour and sift it when the recipe calls for its use. The latter would provide for a greater mass of flour than the former. Maybe have "sifted flour" and "flour" be different ingredients even though they are the same product?
- Some ingredients are `countable` while others are `measurable`.
For example, eggs are countable. They come in discrete units. You can't easily measure out `1/2 egg yolk` but maybe we don't care and let the user decide there. 
- Thoughts on Units (esp. if we want to convert units in the future)
  - Should probably be a reference to another table
  - Should include a "unit system" identifier of some sort. (English units have different sizes depending on what country you're in. e.g., in AU/NZ/USA(nutrition labeling) then a Tsp ~ 5ml, in CA and UK then a Tsp ~ 5.92ml, in conventional US then a Tsp ~ 4.93ml)

Edit:
- Also, probably a field for something like "organic" for an ingredient

**Comment by hay-kot** [comment]
As @dpieski describes. Adding this feature would require a pretty extensive rewrite of the recipe data structure that not adhere to the standard schema, and has a lot of gotchas. That, and I think that it makes the user experience for adding ingredients/recipes more difficult puts me off on including this feature. I have no plans to address this at the moment, but I'm open to a PR or working with another developer who has a good idea on how this can work and keep the user experience "wife approved".

**Comment by phantomtypist** [comment]
My two cents: https://schema.org/recipeIngredient

**Comment by dpieski** [comment]
@phantomtypist yea I've seen that. It's the same schema as Google uses too so it would help with interactions with AI Assistants. 

But that is an API schema, not necessarily how the data has to be stored. You can't do many interesting things with plain text ingredient data.

**Comment by hay-kot** [comment]
Another Update, no real news but... I've come around to the idea that this is something I'd like to support with the caveat that I don't want to change the user experience on the frontend... i.e. it still needs to be very simple to add ingredients and any additional work that gets done needs to be an advanced mode or done via some programming magic. I have 2 schools of thought on how to accomplish this.

### Option 1
Do not change the data structure from string and use a "parsing" method to effectively split the data. For example "1 Cup Flour" would be assuming that the first "word" is the quantity, the second 'word' is the unit and the rest is a description. The benefits of this approach would allow the overall codebase to change very little and keep the user experience effectively unchanged. Only users interested in the advanced features of need to write the structured ingredients, others could write it however they want. 

### Option 2
Completely change how ingredients are stored on the backend and move away from the defined schema. Splitting them into 3 parts, amount, unit, description. This would allow for user completion the frontend and an easily defined structure to work with data and do lots of tasks with the data. The major downside in my opinion is that it turns entering text into data entry which I'd like to avoid if at all possible. anyone can come up with a clever way to simplify the entry of the data I'd be super interested. 

Bottom line is this will happen eventually, but I'm still mulling over the best way to get this done without adding too much complexity

**Comment by phantomtypist** [comment]
I feel like with option 1 and the parser there are six ways to Sunday to shoot yourself in the foot.

**Comment by phantomtypist** [comment]
If you go option 2 (I'm not sure if this is there right now), but you'd need to ask the user on first run if they want US or metric standards..... or come up with a clever way to support both simultaneously.

**Comment by phantomtypist** [comment]
I just thought of something else related to option 2.  What if there is no concept of stored "units".  The database would just hold "amount" and then amount is tied to a base line like teaspoon.  Then in the database we only just store teaspoon quantities as a decimal.  Some baseline to do conversions from and then you don't have to store a unit type.  Then, on the recipe page you just run that through a parser to break "up" (not down, lol) the amount of that base unit.

E.g. so let's say for one ingredient in a receipt it calls for 1 cup.  We allow the person, when creating the recipe, to specify amount "1" and "cup" from a unit drop down list.  Then when the recipe is saved, the unit and amount they specified gets converted/"serialized" to our base unit used in the amounts property.  In this case there would be the number 48 stored for this ingredient because 48 teaspoons equal 1 cup.  Then, when a person pulls up a recipe we can use a parser to convert the 48 teaspoons to something more meaningful to the user like "1 cup".  It'll make it really easy to scale recipes up and down as well.

Another benefit in addition to the scaling of recipes (portions) is that there is no more need to store a "unit" for an ingredient.  The amount value is always stored as some baseline type of unit, e.g. teaspoon in the aforementioned example.

This would also make it really trivial to convert the system between metric and US standard as well.

**Comment by dpieski** [comment]
> I'm still mulling over the best way to get this done without adding too much complexity

@hay-kot, IMHO, I think that the best thing to do would be to keep the recipes working as they are right now for the most part. But we make an addition that, when editing a recipe, you can "convert" the Recipe to an "Advanced Recipe" or something. 

So, current functionality would be maintained. The current tables would not have to be modified.
With an "Advanced Recipe" you would get these ingredient features that have been talked about a lot. 
Your ingredients would be formed of four sections:
- quantity
- unit
- name
- comment
Also shown here:
![Screen Shot 2021-02-20 at 11 54 44 PM](https://user-images.githubusercontent.com/3300522/108616870-258b9680-73d7-11eb-9090-ba6e1145402a.png)

## Option 1
I do not think would be too feasible. I have talked to a friend that is an expert in AI, he said he may play around with training an AI model to parse once I told him how challenging it is. I have accumulated at least a dozen articles and postings regarding how to parse recipes and no one has really been able to get higher than like low 80% accuracy range. I did find a really interesting Thesis from a student at Princeton (she is now at Facebook) where she was using Regex to parse and was able to get over 75% accurate. I will try to find and link her GitHub. I have it bookmarked on a different computer. 

## Option 2
I think the issue here is it would make it harder to manage from parsing - getting everything in the right buckets when you parse in a recipe from a site. (see Sites below).

My proposal is maintaining current functionality for standard Recipes. When you select Edit on a recipe, you have an option to "automatically convert" where we attempt to parse the recipes but have the original next to it so the user can ensure it is parsed correctly. Similarly, when you Add a recipe, you have the option of like changing to an "Advanced Recipe", which, like before, would automatically try to parse any Ingredients already entered, but also allows the user to enter the Ingredients in component form.

## Sites
I do think that, if a plugin model for parsing recipe sites is implemented, a customization can be made in some situations that would allow the recipe to be parsed into an Advanced Recipe, especially if the website provides the recipe with html tags identifying the recipe components, even if it is just partially provided, like quantity, unit, ingredient.

## Other thoughts
- We could prime the Unit list with the unit list from the FDA

**Comment by dpieski** [comment]
@phantomtypist 
> If you go option 2 (I'm not sure if this is there right now), but you'd need to ask the user on first run if they want US or metric standards..... or come up with a clever way to support both simultaneously.

You could default to locale and allow the user to adjust in settings. If locale == en-US, English Units, else SI Units. Maybe conditions for UK, AU, and NZ too?

You wouldn't have to "support" more than one either, you would just have to have conversions setup. I'm sure there is already a library made for this, such as [pint](https://pint.readthedocs.io/en/stable/). Then you could import recipes that are either SI or EN. Parsing wouldn't care about the unit or convert them, it would just put in the db. You would only convert when you view a recipe. 

Consideration, when viewing a recipe, option to switch between English and SI.

Difficult consideration - convert between English and SI (volume) units to SI (weight) units.

**Comment by phantomtypist** [comment]
@dpieski column for "Amount" and another for "UnitType"  UnitTypes as in "volume" and "weight".

**Comment by dpieski** [comment]
> I just thought of something else related to option 2. What if there is no concept of stored "units". The database would just hold "amount" and then amount is tied to a base line like teaspoon. Then in the database we only just store teaspoon quantities as a decimal. 

I do not like this idea. Not everything is easily converted to teaspoons. e.g., a "dash". I really don't see a reason to do this for the most part since it would require a calculation for each time a recipe is viewed instead of simply concatenating the ingredient back together.

Also, not every ingredient would have one of each quantity, unit, name, and comment.

Take, for example, the ingredient "One Apple"
this would parse into
| quantity | unit | ingredient | comment |
|----|----|----|----|
| 1 | | Apple | |

but what about "one clove of garlic"? Which should that be?
The first somewhat seem natural
| quantity | unit | ingredient | comment |
|----|----|----|----|
| 1 | clove  | garlic | |
| 1 | | garlic clove |

But the what about "1 tsp minced garlic"?
Each of the below actually mean two different things. 
| quantity | unit | ingredient | comment |
|----|----|----|----|
| 1 | tsp  | garlic | minced |
| 1 | tsp | minced garlic |

Similarly each of these mean two different things:
"1 cup sifted flour" and
"1 cup flour, sifted"

**Comment by phantomtypist** [comment]
Anybody just know how All Recipes site does it or any other big guy?  Pretty sure this problem has already been solved in places instead of reinventing the wheel.

**Comment by dpieski** [comment]
> column for "Amount" and another for "UnitType" UnitTypes as in "volume" and "weight".

@phantomtypist  I do not think this is really necessary though. You know if it is a "volume" or a "weight" based on the `unit` column. I have a few dozen recipe books at home, and I am not sure if I have ever seen a recipe with a weight that is not in metric - has anyone see that before? Concern here would be Oz being a weight and a volume, so one would have to distinguish Oz and fl-oz.

The concern with converting between Volume and Weight is that it is difficult to do. Densities are not always the same and some of them would just have to be defined based on the comment or have a complex-ingredient where "1 cup flour, sifted' and "1 cup sifted flour" are two different `names` as "flour, sifted" and "sifted flour" would have two different weights based on the differing densities. There is probably a database out there of the top xx number of ingredient-volume-weight conversions. I may look around for it if I get some free time this week.

Similarly, "1 cup sugar" and "1 cup sugar, packed" would have two different weights even though they have the same `volume` - the comment, or complex-ingredient depending on implementation,

**Comment by dpieski** [comment]
> Anybody just know how All Recipes site does it or any other big guy?

I looked up AllRecipes but I can't figure out how to get API access. There is an apps.allrecipes.com but you have to login to go to any of the links.

Here is an interesting [Recipe API article](https://medium.com/groceristar/companies-with-recipe-apis-e9f29a64389c) I found.

## bigoven
I did find `bigoven` API, their ingredients example is, in part:
``` json
"Ingredients": [ 
{ 
"IngredientID": 5247317, 
"DisplayIndex": 0, 
"IsHeading": false, 
"Name": "Soy Sauce", 
"HTMLName": "Soy Sauce", 
"Quantity": 0.333333333333333, 
"DisplayQuantity": "1/3", 
"Unit": "cup", 
"MetricQuantity": 79, 
"MetricDisplayQuantity": "79", 
"MetricUnit": "ml", 
"PreparationNotes": null, 
"IngredientInfo": { 
"Name": "Soy sauce", 
"Department": "Asian", 
"MasterIngredientID": 165, 
"UsuallyOnHand": true 
}, 
"IsLinked": true 
}, 
{ 
"IngredientID": 5247318, 
"DisplayIndex": 1, 
"IsHeading": false, 
"Name": "Dry sherry", 
"HTMLName": "Dry sherry", 
"Quantity": 0.25, 
"DisplayQuantity": "1/4", 
"Unit": "cup", 
"MetricQuantity": 59, 
"MetricDisplayQuantity": "59", 
"MetricUnit": "ml", 
"PreparationNotes": null, 
"IngredientInfo": { 
"Name": "Dry sherry", 
"Department": "Wines", 
"MasterIngredientID": 156, 
"UsuallyOnHand": false 
}, 
"IsLinked": true 
}]
```
## Edamam
`Edamam.com` uses this for ingredients:

Field | Type | Description
-- | -- | --
foodId | string | Food identifier
quantity | float | Quantity of specified measure
measure | Measure | Measure
weight | float | Total weight, g
food | Food | Food
foodCategory | string | Shopping aisle category

##Recipal
[Recipal API](https://www.recipal.com/api-docs#recipe_ingredients) showing the Recipe Ingredient structure.
EXAMPLE OBJECT
``` json 
{
"recipe_ingredient": {
"id": 1204,
"quantity": 1.0,
"unit": "1 cup",
"waste": 0.0,
"ingredient_id": 77,
"recipe_id": 522,
"created_at": "2014-01-12T02:37:16Z",
"updated_at": "2014-01-12T02:37:16Z",
"name": "Butter, salted",
"display_as": "Butter, salted",
"total_grams": 227.0,
"units_and_grams": {
"1 cup": 227.0,
"1 tbsp": 14.2,
"1 pat (1\" sq, 1/3\" high)": 5.0,
"1 stick": 113.0,
"1 lb": 453.592,
"1 gallon": 3632.0,
"1 gram": 1.0,
"1 oz": 28.35
}
}
}
```
##Spoonacular
[Spoonacular API](https://spoonacular.com/food-api/docs#Get-Recipe-Information) showing their recipe information:
``` json
 "extendedIngredients": [
        {
            "aisle": "Milk, Eggs, Other Dairy",
            "amount": 1.0,
            "consitency": "solid",
            "id": 1001,
            "image": "butter-sliced.jpg",
            "measures": {
                "metric": {
                    "amount": 1.0,
                    "unitLong": "Tbsp",
                    "unitShort": "Tbsp"
                },
                "us": {
                    "amount": 1.0,
                    "unitLong": "Tbsp",
                    "unitShort": "Tbsp"
                }
            },
            "meta": [],
            "name": "butter",
            "original": "1 tbsp butter",
            "originalName": "butter",
            "unit": "tbsp"
        },
        {
            "aisle": "Produce",
            "amount": 2.0,
            "consitency": "solid",
            "id": 10011135,
            "image": "cauliflower.jpg",
            "measures": {
                "metric": {
                    "amount": 473.176,
                    "unitLong": "milliliters",
                    "unitShort": "ml"
                },
                "us": {
                    "amount": 2.0,
                    "unitLong": "cups",
                    "unitShort": "cups"
                }
            },
            "meta": [
                "frozen",
                "thawed",
                "cut into bite-sized pieces"
            ],
            "name": "cauliflower florets",
            "original": "about 2 cups frozen cauliflower florets, thawed, cut into bite-sized pieces",
            "originalName": "about frozen cauliflower florets, thawed, cut into bite-sized pieces",
            "unit": "cups"
        }]
```

# Services
Maybe have it setup so users can input an API key for a someone who provides this service?
| Service | Cost | recipes | lines | health labeling | 
|---|---|---|---|---|
[Edamam](https://developer.edamam.com/edamam-nutrition-api) | Free (Devs) | 200/mo | 1000/mo | Basic
[nutritionix](https://www.nutritionix.com/database) | Free (2 users) | - | - | - 

Really Awesome bulk recipe dataset that I am not sure what I want to do with yet... 
http://pic2recipe.csail.mit.edu/
Here is a link to the actual data: https://data.lip6.fr/cadene/im2recipe/

**Comment by hay-kot** [comment]
> So, current functionality would be maintained. The current tables would not have to be modified.
> With an "Advanced Recipe" you would get these ingredient features that have been talked about a lot.
> Your ingredients would be formed of four sections:
> 
> * quantity
> * unit
> * name
> * comment

This is the approach I'm leaning towards. However I think having the split comments would be too confusing. I can't quite think of how you would end up parsing that string to accurately know what is a comment and what is an ingredient. I think have a structure of [Qty, Unit, Ingredient, Comment] Would be the easiest. In fact, I think all I would really need to add is create a "table" editor that breaks down the string into separate parts. Then you can use a separate units/ingredient database table to validate/track what units are being used across that database for autocompletion., similar to how categories and tags work now. I think this allows the most flexibility while also keeping up with the standard schema.

**Comment by dpieski** [comment]
> This is the approach I'm leaning towards. However I think having the split comments would be too confusing. I can't quite think of how you would end up parsing that string to accurately know what is a comment and what is an ingredient. I think have a structure of [Qty, Unit, Ingredient, Comment] Would be the easiest

By "split comments" do you mean like the "softened butter" ingredient in that image? I would agree with you, I think. 
If we have a "convert" option, the we don't really get to choose the format of the ingredient string coming in. The "table editor" idea, even if the parser is only like 70% accurate when converting to an "Advanced" recipe, would be pretty nice. 

I think this [recipe parser from SousChef-backend](https://github.com/mjyuen/SousChef-backend/blob/master/recipeparsing.py) would probably get us most of the way there with a little refactoring. 

If we have a table storing units, it could/should probably be pre-populated and that list above would get most of the way there. There are obviously variations that could be made to many of them still but that can grow for the particular user as you suggested. Maybe account for letter casing if that doesn't. Some exceptions to not caring about cases could be "T." =/= "t." as the first is sometimes used for Tablespoon while the second is sometimes used for Teaspoon. 

For that "table editor" - we could also autocomplete the ingredients too, no? Once a few recipes are loaded up, it could make it easier to enter a new recipe.

**Comment by hay-kot** [comment]
Minor update, I was experimenting with an old npm package I found that may get us most of the way there for supporting this. I've done some work to modernize it and try to make some changes to better suite how mealie works, but overall I was unsuccessful. You can find my attempt [here](https://github.com/hay-kot/mealie/tree/feature/editor-improvements/dev/ingredientScaler) and the original package [here](https://github.com/rviscomi/scale-recipe)

I'm hoping to find someone with some more javascript knowledge to help get this figured out, let me know if anyone interested.

**Comment by h0rnman** [comment]
So, I know nothing of python and its intricacies, but a similar project called Gourmet Recipe Manager does something similar and allows for more 'natural' input of ingredients.  I've been using it for years, but it's not received a new Windows build in a very long time and has no client/server or mobile support, so I'm looking at alternatives.  You might be able to reference some of their code to help with this.

Link is here: https://github.com/thinkle/gourmet/

**Comment by abhchand** [comment]
👍🏾 on this whole idea. It's definitely a more complicated schema but I think scaling ingredients is sort of an "expected" feature of recipe apps.

In  my experience, fully defining an `Ingredient` requires capturing the following components - 

| Key | Value | Description
-- | -- | --
| Quantity | 1/4 | The value of whatever unit |
| Unit | cup | Can be standardized (e.g. tbsp) but could also be free-form (e.g. "cubes" when talking about sugar) |
| Item | Carrots | The actual ingredient name
| Preparation | Finely chopped | How it should be prepared, or in what form it will be used
| Variant | Organic | Optional, but sometimes you want to specify a specific brand, type, or other qualifying label |

Storing these fields lets you -

1. Scale as needed
2. Concatenate all values as a single display string - `1/4 Cup Carrot (Organic), Finely Chopped`

For the `Unit`, the form could be a dropdown or autocomplete of some common unit types (e.g. cups, tbsp, mL, etc..). The end result gets stored as a simple `String` so that they can also populate any custom types (e.g. `cloves` when talking of Garlic). I wouldn't worry a lot about validation on this field, since that greatly expands the scope of this. And we mainly care about isolating `Quantity` since that's what's required for scaling.

**Comment by hay-kot** [comment]
@abhchand, I think your proposal makes sense. The only thing I'd add is an alias so you can merge multiple ingredients on alias in the shopping list. I'd imagine this is a 2-4 day project to get everything all lined up and make sure the import/export work across the board. Given that I'd probably never use any of these features myself, it's not super high on my list. It's probably the last thing I'll get to before we hit 1.0.

I'm happy to help anyone who wants to take this on, otherwise it'll be a while.

**Comment by hay-kot** [comment]
I've Created a task to track progress on this over at https://github.com/hay-kot/mealie/issues/507

I've already started by creating the database layer required and a quick mockup of the routes we'll need to get started. Hopefully this will make it easier for others to jump in an help flesh out all the UI and backend changes we'll need. For coordinating work on this please use the new issue 👍

**Comment by hay-kot** [comment]
Hey all, I've made some significant progress over in the `mealie-next` branch. The next big thing I need to tackle is parsing incoming ingredients. 

I was able to find this [repository](https://github.com/mtlynch/ingredient-phrase-tagger) that does exactly that using CRF++ and some Natural Language Processing. It looks to be extremely powerful and good at correctly parsing ingredients. I was able to get a model trained and create some proof of concept code. What I would like to do is package CRF++ in the docker container and pull down a trained model from somewhere (google drive maybe? ) and include the binary and model in the built container. That way we can easily make subprocess calls to CRF and get the resulting parsed ingredient back. I have this working on my M1 mac using the brew install. 

I'm hoping someone would be willing to take a look at the CI/CD to build in `mealie-next` branch and try to get a working CRF build in the arm and x86 build. If anyone in this thread is up for this, please get in touch and we can work on implementing this. 

IMO this would be a HUGE feature and quality of life improvement for using Mealie, ingredients, and scaling. 

EDIT:

Here's a link to a the trained model https://drive.google.com/file/d/1HUs0dhrLJEbNA3eBsx6PZ9DWfY-oIRtS/view?usp=sharing

**Comment by hay-kot** [comment]
Recipe ingredient scaling is live on https://beta.mealie.io/login

It's half-baked but mostly working. If any of you would be so kind to give it a go and play around with it and let me know what you think of the general UI and process it would be helpful!

**Reply by hay-kot** [reply]
You can manage the foods / units here https://beta.mealie.io/admin/toolbox/foods as well

**Reply by cashpw** [reply]
Comments on the UI ([screenshot](https://www.awesomescreenshot.com/image/14037890?key=20a53a7ec385762ba79849ae33c6f3f4))

1. There was a "Create" button that showed up when I typed in "Kombu" (which was an unknown ingredient) but the same didn't happen for "Water".
1. The text I had typed ("Kombu") was not transferred to the "Create Ingredient" popup "Name" field as I would expect it to be.
1. I had thought that these -/+ icons were for ingredients based on their placement. It took me a little while to identify the "Add" and "Bulk add" buttons.
1. None of these buttons seem to have any hover text. Hover text would be a great place to say something like "Add an ingredient" for the "Add" button to disambiguate it.
1. I thought this "Parse" would be a clickable button, perhaps a toggle-able setting, but it just shows a pop-up and I haven't figured out how to actually _use_ ingredient parsing. Clicking "Add" reveals fields I need to manually set. "Add bulk" reveals a textbox which populates, line by line, the "Notes" fields rather than the appropriate "Unit" or "Food" field.

**Reply by cashpw** [reply]
Regarding point 1 above: The "Create" button appears when I type "Water " (note the trailing space) but not when I type "Water". I consider this a bug but perhaps it is intended?

**Reply by hay-kot** [reply]
Thanks for the feedback! 

1. Create only shows when there are no matches. Obviously, that's not going to work so I'll have to find some other way to display it.
2. I was having trouble getting that working, I agree though. I'll take another look at it.
3/4. Great idea
5.  Parse does work but relies on a CRF++ binary that hasn't yet been installed on the DockerFile. I added the extra dialog to provide some additional context, it doesn't always make the correct prediction and I wanted to a make sure that whoever is doing it should review the changes it suggests. When CRF++ is installed and you click "confirm" a dialog will pop up with the results and you can choose to accept or not.

**Reply by cashpw** [reply]
Just tried out the interface in the [latest v1.0.0 demo](https://github.com/hay-kot/mealie/discussions/645#discussioncomment-1453872) and it's great! The "Add missing ingredient \<foo\>?" button is wonderful!

**Comment by rlglenn** [comment]
I would very much like this feature, and to that point, I have some observations. I would first like ot mention that I don't mind trading a bit of keying to gain more functionality.

1.) Arguably the best way to store unit quantities that may contain fractions is to store the numerator and denominator separately. This could be easily parsed and/or verified and makes scaling precise. I understand this may require a reconsideration of the db schema and routing.

2.) As far as scaling, I've used a IFTTT matrix conversion tables that could be devised per UOM and/or measure type (by weight or volume):

<html>
<body>

  | Gallon | Quart | Pint | Cup | Floz
-- | -- | -- | -- | -- | --
Gallon | 1 | 4 | 8 | 16 | 128
Quart | 4 | 1 | 0.5 | 0.25 | 0.03125
Pint | 8 | 2 | 1 | 0.5 | 0.0625
Cup | 16 | 4 | 2 | 1 | 0.125
Floz | 128 | 32 | 16 | 8 | 1

</body>
</html>

This assumes one would want to change from one UOM to another based on scale or pared-down recipe. First, all of the possible UOM must be defined or able to be added and users restricted to them. All possible conversions need to be calculated beforehand (I would think.) It could all be in one table or separate matrices for scale direction, UOM and/or measure( by weight vs volume - another conversion which would be an even larger task), but in either case, some logic to test if the UOM should remain in the given unit or the next logical UOM (such as the smallest whole numerator after scale applied for example.)  

3.) This would also potentially aid in populating and scaling nutritional values if calculated based solely on ingredients.

**Comment by hay-kot** [comment]
Super excited to say I finally got the Natural Language Processor working in CI/CD and a parser is live and running on [the demo site](https://beta.mealie.io/admin/parser). 

If you want to help improve the parser or submit test-cases you can read up on it here https://nightly.mealie.io/contributors/guides/ingredient-parser/

**Comment by FlexibleToast** [comment]
I know I'm super late to the party and it looks like you already have this underway, but I thought I would throw this out there anyway. This is something that OpenEats did. The project doesn't seem to be maintained and I had troubles trying to get it running when I discovered it years ago. I'm not sure how the user experience was inputting the recipe, but it was a major feature that made me attempt to install OpenEats in the past. 

https://github.com/open-eats

**Reply by hay-kot** [reply]
I think this feature is essentially complete with the exception of some user experience stuff and some backend database modifications. You can preview it at the beta site https://beta.mealie.io/

**Reply by duracell** [reply]
@hay-kot It may be complete for english speaking user, but not for any other language :(
So I think searching for solutions which will work for everyone would be great or did I miss this?
This is the only thing which hold me back from choosing mealie for me, because importing (e.g. tested with chefkoch) works better on tandoor, there the numbers, units and comments are correct imported.
And as far as I saw their is no option for linking units and other stuff, so that if "gram" are imported, I can change this with "Gramm" (german).

**Reply by hay-kot** [reply]
@duracell - You're right that for non-english speakers the parser doesn't work the best. The NLP processor will never work as well because it wasn't trained on a non-english data-set. You can use the brute parser to improve your results, they're not as good as the NLP model but they are helpful sometimes. 

You're correct that there isn't support for alias ingredients. That is something I've thought about but am not actively looking into, as I have other features that I'm more interested in working on at the moment. With one of the most recent builds you can now edit the Foods and Units, so you could edit them or replace them with your primary language and you may get better results. I'm hoping to add the foods and units to be able to be translated on Crowdin so hopefully this process will be automatically done in the future.

If you'd like to see this feature get implemented, I suggest you request it [using the template](https://github.com/hay-kot/mealie/issues/317). We have a pretty wide user-base so someone may want to work on it, or you yourself could work on implementing the feature.

**Reply by duracell** [reply]
@hay-kot Thanks for the answer and explanation :) So training for other languages seems not possible?
And is it possible to change/create the/a parser to get this information? I didn't looked into how it worked there, but tendoor imported everything perfect, at least for chefkoch. I tested the brute parser and it was okay for detecting the number and unit, but not for comments and I also didn't found/understand how to activate this (but just used the beta-demo).

**Reply by hay-kot** [reply]
> So training for other languages seems not possible?

Not really, it's trained on something like a hundred thousand data-samples. I didn't collect the data I just utilized the model so it's unrealistic to get that kind of data for all languages, even just few. 

> And is it possible to change/create the/a parser to get this information? 

When you go to the parser page from a recipe, select Brute Parser from the drop-down and re-run the parser by clicking Parse All

![CleanShot 2022-03-21 at 20 44 21](https://user-images.githubusercontent.com/64056131/159409785-9a4ef746-0898-4fd2-bfd6-fbbd4a2f751a.png)

The parsers are also implemented on the backend as different classes that can be extended. If someone has a better way to do the parser they're welcome to add a PR and I can help get that added in.

**Comment by Goeste** [comment]
This feature would also help to summarize the shopping list... to not have 3 time "1 Apple" but "3 Apples" instead, if I followed this thread correctly, right? That would be a huge benefit for a "planned week" shopping tour.

Best,
goeste

**Comment by timdonovanuk** [comment]
Hi there! There NLP processing feature is superb and working great, now trying to figure out if Mealie can convert between common units, such as ounce to grams? I have looked extensively through various threads but not sure. Thank you!

**Comment by timdonovanuk** [comment]
Just wondering what happened with this feature - Mealie seems to understand imperial units, but only the U.S., Liberia and Myanmar use imperial. It doesn't seem Mealie even understands table/tea spoons?

**Comment by Handkep** [comment]
this feature seems to be dead...

**Reply by michael-genson** [reply]
This has been a feature of Mealie for several years now: https://docs.mealie.io/documentation/getting-started/faq/#how-do-i-enable-smart-ingredient-handling

**Reply by Manny8787** [reply]
How can this be done by default when importing new recipes?

**Reply by FuzzyWereBeast** [reply]
Last option in Household settings.
